src/environment.rs

```diff
@@ -0,0 +1,364 @@
+    fn extend<T: IntoIterator<Item = (K, V)>>(&mut self, iter: T) {
+        // if can mut. borrow current, then we just extend, otherwise it means
+        // it was cloned, and we recreate a new map from iter for current
+        match Rc::get_mut(&mut self.current) {
+            Some(current) => current.extend(iter),
+            None => self.current = Rc::new(HashMap::from_iter(iter)),
+        }
+    }
+}
+
+/// An iterator over the layers of `Environment`.
+///
+/// Created by the [`iter_layers`] method on [`Environment`].
+///
+/// [`iter_layers`]: Environment::iter_layers
+///
+pub struct EnvLayerIter<'a, K: 'a + Hash + Eq, V: 'a + PartialEq> {
+    env: Option<NonNull<Environment<K, V>>>,
+    _marker: PhantomData<&'a Environment<K, V>>,
+}
+
+impl<'a, K: 'a + Hash + Eq, V: 'a + PartialEq> Iterator for EnvLayerIter<'a, K, V> {
+    type Item = Rc<HashMap<K, V>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // SAFETY: NonNull being in an option, we know it cannot be null and can be dereferenceable
+        self.env.map(|env| unsafe {
+            let res = env.as_ref().current.clone();
+            self.env = env
+                .as_ref()
+                .previous
+                .borrow()
+                .as_ref()
+                // SAFETY: can safely create NonNull from Rc
+                .map(|prev| NonNull::new_unchecked(Rc::as_ptr(prev) as *mut _));
+            res
+        })
+    }
+}
+
+/// An iterator over all the elements inside the `Environment`, from the oldest layer to the current one.
+///
+/// Created by the [`iter_elems`] method on [`Environment`].
+///
+pub struct EnvElemIter<'a, K: 'a + Hash + Eq, V: 'a + PartialEq> {
+    env: Vec<NonNull<HashMap<K, V>>>,
+    current_map: std::collections::hash_map::Iter<'a, K, V>,
+}
+
+impl<'a, K: 'a + Hash + Eq, V: 'a + PartialEq> Iterator for EnvElemIter<'a, K, V> {
+    type Item = (&'a K, &'a V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match self.current_map.next() {
+                Some(res) => return Some(res),
+                None => self.current_map = unsafe { self.env.pop()?.as_ref() }.iter(),
+            }
+        }
+    }
+}
+
+/// An iterator over the elements of an `Environment`, from current layer to oldest one.
+/// Keys are guaranteed to appear only once, with actual value.
+///
+/// Created by the [`iter`] method on [`Environment`].
+///
+/// [`iter`]: Environment::iter
+///
+pub struct EnvIter<'a, K: 'a + Hash + Eq, V: 'a + PartialEq> {
+    collapsed_map: hash_map::IntoIter<&'a K, &'a V>,
+}
+
+impl<'a, K: 'a + Hash + Eq, V: 'a + PartialEq> Iterator for EnvIter<'a, K, V> {
+    type Item = (&'a K, &'a V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.collapsed_map.next()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    impl<K: Hash + Eq, V: PartialEq> Environment<K, V> {
+        pub fn deepness(&self) -> usize {
```

Comment:
`depth` is more common I think (deepness is the quality of being deep, while depth is the quantity)